### ISSUE TITLE
Services as dynamic libraries

### ISSUE THREAD

**Issue by DarkEld3r** [exonum/exonum#288]
- [x] Investigate crates for dynamic libraries loading.
- [x] Investigate/discuss changes to `Service` and/or `ServiceFactory` interfaces (traits).
- [ ] Implement required changes.

**Comment by DarkEld3r** [comment]
There are two main crates for dynamic library loading: [libloading](https://crates.io/crates/libloading) and [sharedlib](https://crates.io/crates/sharedlib). The second one provides more library types with different guarantees which is more flexible, but I suppose that we don't need that. Additionally, `libloading` has been updated more recently.

[`dynamic_reload`](https://crates.io/crates/dynamic_reload) provides an interesting possibility of reloading dynamic libraries. Although it is achieved by some tricky machinery of monitoring changes to the specified library-files, so the reloading process is encapsulated in the `dynamic_reload` itself. I suppose we need a more precise control of when and how to reload libraries.

**Comment by DarkEld3r** [comment]
https://github.com/exonum/exonum_sodiumoxide/pull/2
https://github.com/exonum/exonum/pull/295
https://github.com/exonum/exonum-configuration/pull/59

**Comment by DarkEld3r** [comment]
Blocked by https://github.com/rust-lang/cargo/issues/4449.

**Comment by DarkEld3r** [comment]
Won't do. :smile:
